Adding Fullscreen Borderless Window and Resolution Changing to Options Menu

squealingpiggies opened the pull request:
#141
The following code changes the options menu to include a scroll view of resolutions fetched by the love2d engine and a replacement of the fullscreen checkbox with a cycling button to move between windowed, fullscreen (exclusive), and borderless fullscreen window (desktop).

In order to achieve this the following changes were made:

- `OPTIONS.fullscreen` is now a string that hold either `"normal"`, `"fullscreen"`, or `"desktop"`
- `OPTIONS` now contains a `"screen_resolution"` that holds a `width` and `height` that hold the desired screen resolution
- The fullscreen option was changed from a checkbox to a button that cycles through the three settings.
- A scrolllview of buttons containing possible screen resolutions was added bellow the fullscreen option.
- Changed to `set_reference_screen_dimensions` function to keep reference fixed and allow width to stretch or shrink to fit view.
- Added an `updateFullscreen` function to handle updating window and canvas updates on options change.

Windowed with a resolution of 800x600 (1280x720 image)
![Screenshot from 2024-02-06 20-16-33](https://github.com/Calandiel/SongsOfGPL/assets/158529472/4d14c43e-2e03-4a92-81a4-21de0ddf4687)

Fullscreen (exclusive) actually changes the monitor resolution, screenshot is at size of settings value but the monitor will stretch it as one would expect for mismatched aspect ratios.
![Screenshot from 2024-02-06 20-12-20](https://github.com/Calandiel/SongsOfGPL/assets/158529472/e2f809e9-87ec-4112-955c-6dacf8f78d5c)
Borderless Fullscreen Window (desktop) is much sharper than exclusive (unintentional) but still catches the mismatched aspect ratio (intentional).
![Screenshot from 2024-02-06 20-12-42](https://github.com/Calandiel/SongsOfGPL/assets/158529472/59e9ac94-25b9-4943-a85f-e383309df891)

It works well except for 4:3 resolutions where the market can stretch off the edge of the screen.

squealingpiggies committed:
changed fullscreen check box to cycling button

squealingpiggies committed:
added scrollview with buttons for changing screen resolution

squealingpiggies committed:
moved all fullscreen, scaling and canvas update into single function and added to fkey

Calandiel commented:
Thanks! Ill review it in a few hours

Calandiel commented on sote/main.lua:
```diff
@@ -264,3 +273,21 @@ end
 	ui.on_wheelmoved(x, y)
 end
+
+function love.updateFullscreen()
```
I dont believe this is a love2d callback function? The `love` table shouldnt be modified - we can declare it as a standalone function in the scope its needed

squealingpiggies replied:
Would moving the function to the global scope be enough or is there some specific scope/file it you would prefer it be in?

Calandiel replied:
Making a new file for it may be most reasonable. Avoid globals whenever possible

squealingpiggies replied:
What the the design philosophy behind the file hierarchy? Is there some section that I should place to put the update function definition?

Calandiel replied:
Some of these could be better but as they are right now:
```
Inside sote/ which stores the game to allow for convenient `love sote` launching of the game from repos root directory through a terminal:
coa/ - assets for coats of arms
cpml/ - math library, outside of the engine for reasons I don't remember anymore
data/ - "large" game assets and other non code stuff, should be split up and move out of the data directory. iirc wasn't split up to keep licensing info easier to read
engine/ - non GPL and external files
emblems/ - related to coas, different type of graphics
game/ - the game itself
icons/ - open source icons
music/ - open source music, too long to keep in RAM, streamed in
sfx/ - open source sounds, short, kept in RAM
```
Inside `sote/` there's a few "bootup" scripts for the game, there shouldn't be any files being added there and all new game could ought to be within `sote/game/`

As for directories inside `sote/game/`. they mostly divide it by game system.
`climate`, `ecology`, `economy`, etc, are all phases in game's simulation and as such get a directory.
Besides that, there's a `scenes` directory with each subdirectory within corresponding to a concept of a game scene in engines like Unity or Godot (in SotE scenes are basically a FSM).
The `map-modes` directory stores all logic related to coloring tiles in the game world. Not a game system but relatively standalone piece of game code.
`entities` is for "class definitions" of various data types stored on the game world.
`raws` is for static data definitions that are loaded before the game itself starts. Corresponds to `raws` in Dwarf Fortress or the `common` folder in games by Paradox Interactive. Things like "biomes", "races", "events", etc.
Lastly, `ai`, I think is quite self expanatory.

As for the fullscreen function, keeping it on the options object should do the trick

Calandiel commented on sote/game/options.lua:
```diff
@@ -7,7 +7,7 @@ local opt = {}
 	return {
 		["volume"] = 0,
-		["fullscreen"] = true,
+		["fullscreen"] ="normal",
```
Standalone strings arent type safe for de facto enums, it needs a type declaration for the LSP

squealingpiggies replied:
I don't know exactly what you want here as I have only been working with lua for a week now and only understand it through your code and quick web searches. Is this just annotation for the language server on the function or variable? Some kind of reference or example would be helpful.
Edit:
I think I get what you mean now but to be clear, you want some code that defines the strings inside a global table to be used like an enum of strings like
```
---@enum FULLSCREEN
FULLSCREEN = {
  NORMAL = "normal",
  EXCLUSIVE = "exclusive",
  DESKTOP = "desktop"
}
```
and to replace all instances of the strings with references to this variable.

Calandiel replied:
Ah, sorry, I should have been more specific.

Wrt your edit, thats one way to do it, yeah. 
Though, I think in SotE the most common way is with aliases as enums as that avoids having to hope for the JIT to inline table lookup:
```lua
--@alias enumname value_a | value_b | value_c
```

You'd then need to annotate the function to mark the parameter of the type as `enumname`. See `function Rect:subrect` in `sote/engine/ui.lua` for an example of annotating a function with parameter types (in particular, enums).

SotE uses this LSP: https://github.com/LuaLS/lua-language-server
Its highly recommended to get an extension for it for your IDE

Calandiel commented:
BTW, @ineveraskedforthis 
This MR also changes the screen scaling behavior to make the game scretch on devices that dont fit the target resolution.
This is a design change more than a fix, we should discuss it before deciding if its a desired behavior.
What are your thoughts?

ineveraskedforthis commented:
> BTW, @ineveraskedforthis This MR also changes the screen scaling behavior to make the game scretch on devices that dont fit the target resolution. This is a design change more than a fix, we should discuss it before deciding if its a desired behavior. What are your thoughts?
As long as it doesn't require any maintenance and things work as usual on "main" resolution, I am fine with such change. Having some way to handle multiple resolutions without maintenance cost is better than not having support for multiple resolutions at all.

squealingpiggies committed:
moved updateFullscreen from love table to global function

squealingpiggies committed:
replaced OPTIONS.fullscreen with enum like variable holding defined strings

squealingpiggies committed:
moved FULLSCREEN to OPTIONS file

squealingpiggies committed:
missed a few standalone strings

Calandiel commented:
> > BTW, @ineveraskedforthis This MR also changes the screen scaling behavior to make the game scretch on devices that dont fit the target resolution. This is a design change more than a fix, we should discuss it before deciding if its a desired behavior. What are your thoughts?
> 
> As long as it doesn't require any maintenance and things work as usual on "main" resolution, I am fine with such change. Having some way to handle multiple resolutions without maintenance cost is better than not having support for multiple resolutions at all.

Maybe it should be a separate type of a setting. "fullscreen (stretch)" and "fullscreen (keep aspect)"

Wdyt? @ineveraskedforthis @squealingpiggies

squealingpiggies commented:
From my understanding as a user, most programs graphics end up stretched if you set the screen resolution to a different aspect ratio than the monitor on fullscreen thus I was trying to intentionally get that from my code. But from the coding side, love2d has two fullscreen modes as options from one version of `setFullscreen` that takes boolean and string as option. The options only works if the boolean is true and, from my experimentation, they do either one of two things: set the monitor resolution to the window size (exclusive) or stretch the window cover the whole monitor (desktop). I simply implemented some checks to pass one of those two options.
If preventing screen stretching is desired, having a keep aspect option could easily be implemented as another options variable with a checkbox and checking it in the `UpdateFullscreen`. Alternatively, screen resolution and fullscreen options could be made mutually exclusive, where the exclusive and desktop will always use the monitor resolutions, so the fullscreen never stretches the drawing.

squealingpiggies committed:
removed extraneous print statements

Calandiel commented:
(Just a heads up, I'm in Tokyo so timezones make this a bit slower than it needs to be - I'll take a look again in circa 10h)

squealingpiggies committed:
changed enum to alias and moved updateFullscreen to options file

squealingpiggies committed:
added alias for love.alignMode to clear warnings from main and options files

squealingpiggies committed:
added fitscreen checkbox to options that matches scaling factor to monitor aspect ratio when fullscreen

squealingpiggies commented:
Now has a new fitscreen variable in the `OPTIONS` linked to a checkbox in the options menu. This variable is used to match fullscreen stretching to monitor aspect ratio but still f4 to a different ratio window, if desired.
The new variable, if set to true, will prevent width stretching/squishing when going fullscreen even if the aspect ratio is different from the monitor. This allows the user to have a custom sized window and use F4 to toggle between exclusive and windowed mode while keeping all square images properly square.
I also figured out how to fit the alias but had to annotate an `Options` class that used the `Fullscreen` alias and changed `updateFullscreen` to take the alias as a parameter.
I moved the update function to opt in _game.options.lua_ since it is tied heavily to the annotated `Options` class. Additionally, I annotated the returns for `opt.init` and `opt.load` and created an alias for `love.AlignMode` in _engine.ui.lua_ to clean up the last remaining warnings on the three files touched (_engine.ui.lua_ still has more warnings to solve).

squealingpiggies committed:
fixed F4 require calls

Calandiel commented on sote/game/options.lua:
```diff
@@ -1,13 +1,28 @@
+---@feild ["exploration"] number
+---@feild ["travel-start"] number
+---@feild ["travel-end"] number
+---@feild ["screen_resolution"] {width number, height number}
```
I presume it's meant to be `field`?

Calandiel commented:
> From my understanding as a user, most programs graphics end up stretched if you set the screen resolution to a different aspect ratio than the monitor on fullscreen thus I was trying to intentionally get that from my code. But from the coding side, love2d has two fullscreen modes as options from one version of `setFullscreen` that takes boolean and string as option. The options only works if the boolean is true and, from my experimentation, they do either one of two things: set the monitor resolution to the window size (exclusive) or stretch the window cover the whole monitor (desktop). I simply implemented some checks to pass one of those two options. If preventing screen stretching is desired, having a keep aspect option could easily be implemented as another options variable with a checkbox and checking it in the `UpdateFullscreen`. Alternatively, screen resolution and fullscreen options could be made mutually exclusive, where the exclusive and desktop will always use the monitor resolutions, so the fullscreen never stretches the drawing.

Most programs (well, well made ones) don't *stretch* the resolution, they expand the rendered view and adapt the sizes of rendered ui widgets (relative to the whole screen). See Skyrim, Europa Universalis IV, Factorio, Godot or Paint for examples.
You can also see Unity or Godot for automatically scaling UIs, search terms would be things like "anchors", "keep aspect", "keep width", etc.
SotE's UI library doesn't have support for this - if you stretch resolution square UI widgets stop being squares. That behavior was popular in very old games such as original releases of Gothic or Heroes of Might and Magic. That was due to them not taking into account that in the future the average user's aspect ratio may change.

I think it's fine to have it but it ought to be spelled out loudly that the option will *actually* stretch the screen ^-^

Calandiel commented on sote/game/options.lua:
```diff
@@ -1,13 +1,28 @@
 		["volume"] = 0,
-		["fullscreen"] = true,
+		["fullscreen"] = "false",
+		["fitscreen"] = true,
```
While this works, bitser (used for serialization in sote), uses love's built-in filesystem to serialize/deserialize objects.
Currently, if someone has a file with the previous version of the program, bitser will load the previous versions 

We ought to check for a version and if it's missing, reinitialize the options object.
Something like the following:
```lua
function opt.init()
    return {
        ["version"] = "v0.3.1"
        -- etc
```

Then, it'd need to be checked in the verify method:
```lua
function opt.verify()
	local default = opt.init()
        if OPTIONS.version ~= default.version then
            OPTIONS = default
            return
        end

	for i, j in pairs(default) do
		if OPTIONS[i] == nil then
			OPTIONS[i] = j
		end
	end
end
```
Something along those lines will ensure that users with older versions of the options schema will still be able to boot the game

Calandiel commented:
Other than that, seems sensible, we should merge it in

squealingpiggies committed:
fixed annontations and added version variable to be checked in load function

squealingpiggies commented:
Added the version field to options and fixed my typo. Also added argument passing to the bash script.

Calandiel commented:
Seems solid. Thanks!
